File: campaign/src/issue-fixer.ts
import type { Env } from "./index";

const MAX_PER_RUN   = 3;
const AUTOFIX_LABEL = "autofix-attempted";

// Mappar deplojat filnamn → källfil i repot
const FILE_MAP: Record<string, string> = {
  "app.js":    "app/public/app.js",
  "i18n.js":   "app/public/i18n.js",
  "style.css": "app/public/style.css",
  "index.ts":  "app/src/index.ts",
};

interface GithubIssue {
  number: number;
  title: string;
  body: string;
  labels: Array<{ name: string }>;
}

async function ghApi(token: string, path: string, method = "GET", body?: unknown): Promise<unknown> {
  const resp = await fetch(`https://api.github.com${path}`, {
    method,
    headers: {
      Authorization: `Bearer ${token}`,
      Accept: "application/vnd.github+json",
      "User-Agent": "politiker-webapp-campaign",
      "Content-Type": "application/json",
    },
    body: body ? JSON.stringify(body) : undefined,
  });
  return resp.json();
}

async function callClaude(apiKey: string, prompt: string): Promise<string> {
  const resp = await fetch("https://api.anthropic.com/v1/messages", {
    method: "POST",
    headers: { "x-api-key": apiKey, "anthropic-version": "2023-06-01", "content-type": "application/json" },
    body: JSON.stringify({
      model: "claude-sonnet-4-6",
      max_tokens: 4000,
      messages: [{ role: "user", content: prompt }],
    }),
  });
  const data = await resp.json() as { content: Array<{ text: string }> };
  return data.content[0].text.trim();
}

function parseStackTrace(body: string): { file: string; line: number } | null {
  const m = body.match(/(\w+\.(?:js|ts|css)):(\d+)/);
  if (!m) return null;
  return { file: m[1], line: parseInt(m[2]) };
}

function parseErrorMessage(body: string): string {
  const m = body.match(/\[Auto-rapport\] ([^\n]+)/);
  return m?.[1] ?? body.slice(0, 200);
}

export async function runIssueFixer(env: Env): Promise<void> {
  const repo = env.GITHUB_REPO;
  const issues = await ghApi(env.GITHUB_FEEDBACK_TOKEN,
    `/repos/${repo}/issues?labels=user-reported&state=open&per_page=20`
  ) as GithubIssue[];

  const toFix = issues.filter(i =>
    !i.labels.some(l => l.name === AUTOFIX_LABEL)
  ).slice(0, MAX_PER_RUN);

  if (!toFix.length) { console.log("issue-fixer: inga nya issues att åtgärda"); return; }

  // Hämta HEAD-SHA för main
  const mainRef = await ghApi(env.GITHUB_FEEDBACK_TOKEN, `/repos/${repo}/git/ref/heads/main`) as { object: { sha: string } };
  const mainSha = mainRef.object.sha;

  for (const issue of toFix) {
    const errorMsg = parseErrorMessage(issue.body ?? "");
    const stack    = parseStackTrace(issue.body ?? "");
    const sourcePath = stack ? FILE_MAP[stack.file] : null;

    // Markera som under bearbetning
    await ghApi(env.GITHUB_FEEDBACK_TOKEN, `/repos/${repo}/issues/${issue.number}/labels`, "POST",
      { labels: [AUTOFIX_LABEL] }
    );

    if (!sourcePath) {
      console.log(`issue-fixer: #${issue.number} — kan inte mappa stack-trace, hoppar över`);
      continue;
    }

    try {
      // Hämta källfil från GitHub
      const fileData = await ghApi(env.GITHUB_FEEDBACK_TOKEN,
        `/repos/${repo}/contents/${sourcePath}?ref=main`
      ) as { content: string; sha: string };

      const content = atob(fileData.content.replace(/\n/g, ""));
      const lines   = content.split("\n");
      const lineNum = stack!.line;
      const context = lines.slice(Math.max(0, lineNum - 30), lineNum + 30).join("\n");

      // Be Claude generera fix
      const fixPrompt = `Du är en kodfix-agent för politiker.denied.se.

Fel rapporterat i produktion (issue #${issue.number}):
Felmeddelande: ${errorMsg}
Fil: ${sourcePath}, rad ${lineNum}

Kod runt felet (rad ${Math.max(1, lineNum - 30)}–${lineNum + 30}):
\`\`\`
${context}
\`\`\`

Fullständigt issue-innehåll:
${(issue.body ?? "").slice(0, 1000)}

Returnera ENBART den fullständiga fixade filen (hela innehållet, inte bara ändringen).
Om du inte kan göra en säker fix, svara exakt: INGEN FIX`;

      const fixedContent = await callClaude(env.ANTHROPIC_API_KEY, fixPrompt);

      if (fixedContent.startsWith("INGEN FIX")) {
        console.log(`issue-fixer: #${issue.number} — Claude kunde inte hitta fix`);
        continue;
      }

      // Skapa branch
      const branch = `claude/autofix-${issue.number}`;
      await ghApi(env.GITHUB_FEEDBACK_TOKEN, `/repos/${repo}/git/refs`, "POST", {
        ref: `refs/heads/${branch}`, sha: mainSha,
      });

      // Commita fix
      await ghApi(env.GITHUB_FEEDBACK_TOKEN, `/repos/${repo}/contents/${sourcePath}`, "PUT", {
        message: `Fix: autofix av issue #${issue.number} — ${errorMsg.slice(0, 60)}`,
        content: btoa(unescape(encodeURIComponent(fixedContent))),
        sha: fileData.sha,
        branch,
      });

      // Öppna PR
      await ghApi(env.GITHUB_FEEDBACK_TOKEN, `/repos/${repo}/pulls`, "POST", {
        title: `Autofix issue #${issue.number}: ${errorMsg.slice(0, 50)}`,
        body: `Automatisk fix av https://github.com/${repo}/issues/${issue.number}\n\nGenererad av issue-fixer Worker.`,
        head: branch,
        base: "main",
      });

      console.log(`issue-fixer: PR skapad för issue #${issue.number}`);
    } catch (e) {
      console.error(`issue-fixer: fel vid fix av #${issue.number}:`, e);
    }
  }
}

File: infra/issue-fixer.py
#!/usr/bin/env python3
"""
Autonom issue-fixer för politiker-webapp.
Läser Gmail IMAP för nya feedback-mail, hämtar GitHub-issue-kontext,
kör Claude CLI för att analysera och fixa buggen, skapar PR.
Körs av systemd-timer (issue-fixer.timer) dagligen kl 11:00.
Max 3 issues per körning.
"""
import imaplib, email, re, os, sys, json, subprocess, logging, urllib.request
from email.header import decode_header

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s issue-fixer %(levelname)s %(message)s",
    handlers=[logging.StreamHandler(sys.stdout)],
)
log = logging.getLogger()

ENV_FILE      = os.path.expanduser("~/.appdata/.config/.env")
REPO_PATH     = "/home/berduf/GitHub/politiker-webapp"
GITHUB_REPO   = "blixten85/politiker-webapp"
MAIL_SUBJECT  = "Ny feedback"
MAX_PER_RUN   = 3
AUTOFIX_LABEL = "autofix-attempted"

def load_env():
    env = {}
    if os.path.exists(ENV_FILE):
        with open(ENV_FILE) as f:
            for line in f:
                line = line.strip()
                if "=" in line and not line.startswith("#"):
                    k, _, v = line.partition("=")
                    env[k.strip()] = v.strip().strip('"').strip("'")
    return env

def decode_str(s):
    parts = decode_header(s or "")
    result = []
    for part, charset in parts:
        if isinstance(part, bytes):
            result.append(part.decode(charset or "utf-8", errors="replace"))
        else:
            result.append(part)
    return "".join(result)

def fetch_issue_body(issue_url):
    # Konvertera https://github.com/owner/repo/issues/123 → API-URL
    m = re.search(r"github\.com/([^/]+/[^/]+)/issues/(\d+)", issue_url)
    if not m:
        return None, None
    repo, num = m.group(1), m.group(2)
    api_url = f"https://api.github.com/repos/{repo}/issues/{num}"
    req = urllib.request.Request(
        api_url,
        headers={"Accept": "application/vnd.github+json", "User-Agent": "politiker-issue-fixer"},
    )
    with urllib.request.urlopen(req, timeout=15) as r:
        data = json.loads(r.read())
    labels = [l["name"] for l in data.get("labels", [])]
    return data.get("body", ""), labels, int(num)

def add_label(issue_num, label):
    subprocess.run(
        ["gh", "issue", "edit", str(issue_num), "--repo", GITHUB_REPO, "--add-label", label],
        capture_output=True, timeout=30,
    )

def run_claude_fix(issue_num, issue_body):
    prompt = f"""Du är en autonom kodfix-agent för politiker.denied.se.
Repot finns på: {REPO_PATH}
Branch: kör alltid från senaste main

En produktionsbugg (GitHub issue #{issue_num}) har rapporterats:

---
{issue_body[:3000]}
---

GÖR FÖLJANDE:
1. cd {REPO_PATH} && git fetch origin && git checkout -b claude/autofix-{issue_num} origin/main
2. Läs relevant fil baserat på stack-trace eller felbeskrivning
3. Hitta rotorsaken och gör minimal fix (ändra BARA vad som krävs)
4. git add <ändrade filer> && git commit -m "Fix: autofix av issue #{issue_num}"
5. git push origin claude/autofix-{issue_num}
6. gh pr create --base main --head claude/autofix-{issue_num} --title "Autofix issue #{issue_num}" --body "Automatisk fix av https://github.com/{GITHUB_REPO}/issues/{issue_num}"

Om du inte kan hitta en säker fix — skriv "INGEN FIX" och gör ingenting.
Ingen refaktorisering. Inga extra ändringar."""

    log.info("Kör Claude för issue #%d", issue_num)
    result = subprocess.run(
        ["claude", "--dangerously-skip-permissions", "-p", prompt],
        cwd=REPO_PATH,
        capture_output=True,
        text=True,
        timeout=300,
        env={**os.environ, "ANTHROPIC_API_KEY": load_env().get("ANTHROPIC_API_KEY", "")},
    )
    output = result.stdout + result.stderr
    log.info("Claude output (issue #%d): %s", issue_num, output[:500])
    return "INGEN FIX" not in output

def main():
    env = load_env()
    imap_user = env.get("GMAIL_EMAIL")
    imap_pw   = env.get("GMAIL_PASSWORD")
    if not imap_user or not imap_pw:
        log.error("GMAIL_EMAIL / GMAIL_PASSWORD saknas"); sys.exit(1)

    mail = imaplib.IMAP4_SSL("imap.gmail.com", 993)
    mail.login(imap_user, imap_pw)
    mail.select("INBOX")

    _, data = mail.search(None, f'(UNSEEN SUBJECT "{MAIL_SUBJECT}")')
    if not data or not data[0]:
        log.info("Inga nya feedback-mail")
        mail.logout()
        return

    seqnums = data[0].split()
    log.info("%d nya feedback-mail", len(seqnums))
    processed = 0

    for seq in seqnums:
        if processed >= MAX_PER_RUN:
            log.info("Max %d issues per körning nådd", MAX_PER_RUN)
            break

        _, fdata = mail.fetch(seq, "(RFC822)")
        raw = next((i[1] for i in fdata if isinstance(i, tuple) and isinstance(i[1], bytes)), None)
        if not raw:
            continue

        msg = email.message_from_bytes(raw)
        subject = decode_str(msg.get("Subject", ""))

        # Extrahera GitHub-issue-URL från mail-body
        body_text = ""
        if msg.is_multipart():
            for part in msg.walk():
                if part.get_content_type() in ("text/plain", "text/html"):
                    body_text += part.get_payload(decode=True).decode(errors="replace")
        else:
            body_text = msg.get_payload(decode=True).decode(errors="replace")

        issue_urls = re.findall(r"https://github\.com/[^\s\"<>]+/issues/\d+", body_text)
        if not issue_urls:
            log.info("Inget GitHub-issue-URL hittades i mail: %s", subject)
            mail.store(seq, "+FLAGS", "\\Seen")
            continue

        issue_url = issue_urls[0]
        log.info("Bearbetar issue: %s", issue_url)

        try:
            issue_body, labels, issue_num = fetch_issue_body(issue_url)
        except Exception as e:
            log.warning("Kunde inte hämta issue: %s", e)
            mail.store(seq, "+FLAGS", "\\Seen")
            continue

        if AUTOFIX_LABEL in (labels or []):
            log.info("Issue #%d redan bearbetad, hoppar över", issue_num)
            mail.store(seq, "+FLAGS", "\\Seen")
            continue

        # Markera issue som under bearbetning innan Claude körs
        add_label(issue_num, AUTOFIX_LABEL)

        success = run_claude_fix(issue_num, issue_body)
        if success:
            log.info("Fix skapad för issue #%d", issue_num)
        else:
            log.info("Ingen fix möjlig för issue #%d", issue_num)

        mail.store(seq, "+FLAGS", "\\Seen")
        processed += 1

    mail.logout()
    log.info("Klart: %d issues bearbetade", processed)

if __name__ == "__main__":
    main()
